File: BaekJoon/2110(공유기설치).py
#https://www.acmicpc.net/problem/2110

import sys
input = sys.stdin.readline
n,c = map(int,input().split())
x = list()
for _ in range(n):
    x.append(int(input()))
x = sorted(x)
start = x[1]-x[0]
end = x[-1]-x[0]
result = 0
while start<=end :
    mid = (start+end)//2
    value = x[0]
    cnt = 1
    for i in range(1,len(x)):
        if(mid <= x[i]-value):
            value = x[i]
            cnt +=1
    if(cnt>=c):
        start = mid+1
        result = mid
    else:
        end = mid -1
print(result)
        

File: BOJ/javascript/2110(공유기 설치).js
// https://www.acmicpc.net/problem/2110
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(filePath).toString().split("\n");
let coord = [];
const [N, C] = input[0].split(" ").map((el) => +el);
for (let i = 1; i < input.length; i++) {
  coord.push(+input[i]);
}
coord.sort((a, b) => a - b);
function count(arr, dist) {
  let prev = arr[0];
  let cnt = 1;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] - prev >= dist) {
      cnt++;
      prev = arr[i];
    }
  }
  return cnt;
}
function solution() {
  let answer = 0;
  let start = 1;
  let end = coord[coord.length - 1];
  while (start <= end) {
    let mid = Math.floor((start + end) / 2);

    if (count(coord, mid) >= C) {
      start = mid + 1;
      answer = mid;
    } else {
      end = mid - 1;
    }
  }
  return answer;
}
console.log(solution());
